src/adapter/src/session/vars.rs

```diff
@@ -1887,16 +1957,37 @@ impl Value for Vec<String> {
 impl Value for Option<String> {
     const TYPE_NAME: &'static str = "optional string";
 
-    fn parse(s: &str) -> Result<Self::Owned, ()> {
+    fn parse(s: &str) -> Result<Option<String>, ()> {
+        match s {
+            "" => Ok(None),
+            _ => <str as Value>::parse(s).map(Some),
+        }
+    }
+
+    fn format(&self) -> String {
+        match self {
+            Some(s) => s.format(),
+            None => "".into(),
+        }
+    }
+}
+
+impl<T: Value + Clone> Value for Option<T>
+where
+    T: Value + Clone + ToOwned<Owned = T>,
+{
+    const TYPE_NAME: &'static str = T::TYPE_NAME;
+
+    fn parse(s: &str) -> Result<Option<T>, ()> {
         match s {
             "" => Ok(None),
-            _ => Ok(Some(s.into())),
+            _ => T::parse(s).map(Some),
         }
     }
 
     fn format(&self) -> String {
         match self {
-            Some(s) => s.clone(),
+            Some(v) => v.format(),
             None => "".into(),
         }
     }
```

Comment:
It would be nice to just have a singe `Value` impl for `Option`s wrapping another `Value` and get rid of the special impl for `Option<String>`. Unfortunately, `String` is not a `Value` (`str` is), and I couldn't get Rust to accept any type bounds using `T::Owned` I tried.